File: app/src/main/java/com/fyp/agrifarm/app/weather/model/WeatherRepository.java
package com.fyp.agrifarm.app.weather.model;

import android.app.Application;
import android.content.SharedPreferences;
import android.util.Log;
import android.widget.Toast;

import androidx.lifecycle.MutableLiveData;
import androidx.preference.PreferenceManager;

import com.google.type.LatLng;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Locale;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

import static com.fyp.agrifarm.app.ExtensionsKt.getSharedPrefs;

public class WeatherRepository {

    private static WeatherRepository instance;
    private static final String APP_ID = "e1f236cb66f5933607fd7905b42b39c9";
    private static final String BASE_URL = "https://api.openweathermap.org/";
    private List<HourlyObject> hourlyObject = new ArrayList<>();
    public MutableLiveData<List<HourlyObject>> hourlyList = new MutableLiveData<>();
    private List<DailyObject> dailyObject = new ArrayList<>();
    public MutableLiveData<List<DailyObject>> dailyList = new MutableLiveData<>();
    public MutableLiveData<CurrentWeatherObject> currentWeather = new MutableLiveData<>();

    public WeatherRepository getInstance() {
        if (instance == null) {
            instance = new WeatherRepository();
        }
        return instance;
    }

    private String getWeatherUnit(Application application) {
        SharedPreferences sharedPref = PreferenceManager.getDefaultSharedPreferences(application.getApplicationContext());
        String WeatherPref = sharedPref.getString("weatherUnit", "-1");
        return WeatherPref;
    }

    private LatLng getLocation(Application application) {
        String latitude = getSharedPrefs(application.getApplicationContext()).getString("lat", null);
        String longitude = getSharedPrefs(application.getApplicationContext()).getString("lon", null);
        try {
            double lat = Double.parseDouble(latitude);
            double lng = Double.parseDouble(longitude);
            return LatLng.newBuilder().setLatitude(lat).setLongitude(lng).build();
        } catch (Exception e) {
            return null;
        }
    }

    public void getHourlyForecastByApi(Application application) {

        String unit = getWeatherUnit(application);
        String lat = null;
        String lon = null;

        LatLng location = getLocation(application);
        if (location != null) {
            lat = String.valueOf(location.getLatitude());
            lon = String.valueOf(location.getLongitude());
        }

        if (unit.equals("Celsius")) {
            unit = "metric";
        } else if (unit.equals("imperial")) {
            unit = "imperial";

        }

        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .build();
        WeatherService service = retrofit.create(WeatherService.class);
        Call<WeatherHourlyResponse> call = service.getHourlyForecastList(lat, lon, unit, "current", APP_ID);
        call.enqueue(new Callback<WeatherHourlyResponse>() {
            @Override
            public void onResponse(Call<WeatherHourlyResponse> call, Response<WeatherHourlyResponse> response) {
                if (response.isSuccessful()) {
                    WeatherHourlyResponse weatherHourlyResponse = response.body();
                    hourlyObject = weatherHourlyResponse.weatherHourly;
                    hourlyList.postValue(hourlyObject);
                } else {
                    //TODO Implement LocalCache
                }
            }

            @Override
            public void onFailure(Call<WeatherHourlyResponse> call, Throwable t) {
                Toast.makeText(application, "Failed", Toast.LENGTH_SHORT).show();
            }
        });
    }

    public void getDailyListForecastByApi(Application application) {

        String unit = getWeatherUnit(application);
        String lat = null;
        String lon = null;

        LatLng location = getLocation(application);
        if (location != null) {
            lat = String.valueOf(location.getLatitude());
            lon = String.valueOf(location.getLongitude());
        }

        if (unit.equals("Celsius")) {
            unit = "metric";
        } else {
            unit = "imperial";

        }
        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .build();
        WeatherService service = retrofit.create(WeatherService.class);
        Call<WeatherHourlyResponse> call = service.getDailyForecastList(lat, lon, unit, "current", APP_ID);
        call.enqueue(new Callback<WeatherHourlyResponse>() {
            @Override
            public void onResponse(Call<WeatherHourlyResponse> call, Response<WeatherHourlyResponse> response) {
                if (response.isSuccessful()) {
                    WeatherHourlyResponse weatherHourlyResponse = response.body();
                    dailyObject = weatherHourlyResponse.weatherDaily;
                    dailyList.postValue(dailyObject);
                } else {
                    //TODO Implement LocalCache
                }
            }

            @Override
            public void onFailure(Call<WeatherHourlyResponse> call, Throwable t) {
                Toast.makeText(application, "Failed", Toast.LENGTH_SHORT).show();
            }
        });
    }

    public void FetchDailyForecast(Application application) {
        String unit = getWeatherUnit(application);
        String lat = null;
        String lon = null;

        LatLng location = getLocation(application);
        if (location != null) {
            lat = String.valueOf(location.getLatitude());
            lon = String.valueOf(location.getLongitude());
        }

        if (unit.equals("Celsius")) {
            unit = "metric";
        } else {
            unit = "imperial";

        }
        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .build();
        WeatherService service = retrofit.create(WeatherService.class);
        Call<WeatherDailyResponse> call = service.getCurrentWeatherData(lat, lon, APP_ID, unit);
        call.enqueue(new Callback<WeatherDailyResponse>() {
            @Override
            public void onResponse(Call<WeatherDailyResponse> call, Response<WeatherDailyResponse> response) {
                if (response.isSuccessful()) {
                    WeatherDailyResponse weatherDailyResponse = response.body();
//                Log.d("FSDSF","" + response.body().main.temp);
                    Weather weather = weatherDailyResponse.weather.get(0);
                    SimpleDateFormat sdf = new SimpleDateFormat("EEEE", Locale.ENGLISH);
                    Date d = new Date();
                    String dayOfTheWeek = sdf.format(d);
                    String description = weather.description;
                    String tempertaure = String.valueOf(weatherDailyResponse.main.temp);
                    tempertaure = tempertaure.substring(0, 2);
                    String humidity = String.valueOf(weatherDailyResponse.main.humidity);
                    String icon = weather.icon;
                    double windvalue = weatherDailyResponse.wind.speed * 3.6;
                    String windpressure = String.valueOf(windvalue);
                    windpressure = windpressure.substring(0, 3);
                    CurrentWeatherObject currentWeatherObject = new CurrentWeatherObject(dayOfTheWeek,
                            tempertaure, description, humidity, windpressure, icon);
                    currentWeather.postValue(currentWeatherObject);
                }
                else
                {
                    //TODO Implement This
                }

            }

            @Override
            public void onFailure(Call<WeatherDailyResponse> call, Throwable t) {
                Log.d("HALA", t.getMessage());
            }
        });

    }


}

File: app/src/main/java/com/fyp/agrifarm/app/weather/repo/WeatherRepository.kt
package com.fyp.agrifarm.app.weather.repo

import android.app.Application
import androidx.lifecycle.MutableLiveData
import androidx.preference.PreferenceManager
import com.fyp.agrifarm.app.log
import com.fyp.agrifarm.app.sharedPrefs
import com.fyp.agrifarm.app.toast
import com.fyp.agrifarm.app.weather.IconMapsUtil
import com.fyp.agrifarm.app.weather.model.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.text.SimpleDateFormat
import java.util.*


class WeatherRepository private constructor() {


    companion object {
        private val instance: WeatherRepository = WeatherRepository()
        private const val APP_ID = "e1f236cb66f5933607fd7905b42b39c9"
        private const val BASE_URL = "https://api.openweathermap.org/"
        private const val IMPERIAL = "imperial"
        private const val METRIC = "metric"

        fun getInstance(app: Application): WeatherRepository {
            instance.application = app
            return instance
        }
    }

    private lateinit var application: Application
    private val retrofit: Retrofit = Retrofit.Builder()
            .baseUrl(BASE_URL)
            .addConverterFactory(GsonConverterFactory.create())
            .build()
    private val service = retrofit.create(WeatherService::class.java)
    private val dayFormat = SimpleDateFormat("EEEE", Locale.ENGLISH)

    var hourlyList = MutableLiveData<List<HourlyWeatherItem>>()
    var dailyList = MutableLiveData<List<DailyWeatherItem>>()
    var currentWeather = MutableLiveData<CurrentWeatherItem>()

    private var weatherUnit: String? = null

    private var lat: String? = null
    private var lon: String? = null

    fun setWeatherUnit(newUnit: String?) {
        if (newUnit == null) return
        weatherUnit = if (newUnit.contentEquals("Fahrenheit")) IMPERIAL else METRIC
    }

    private fun getWeatherUnit() {
        if (weatherUnit == null) {
            val sharedPref = PreferenceManager.getDefaultSharedPreferences(application.applicationContext)
            val st = sharedPref.getString("weatherUnit", "-1")
            setWeatherUnit(st)
        }
    }

    private val location: Unit
        get() {
            log("$lat $lon")
            lat = application.applicationContext.sharedPrefs.getString("lat", null)
            lon = application.applicationContext.sharedPrefs.getString("lon", null)
        }

    private fun preprocess() {
        getWeatherUnit()
        location
    }

    //TODO Implement LocalCache
    fun fetchWeather() {
        preprocess()
        val call = service.fetchWeather(lat, lon, weatherUnit, "minutely", APP_ID)
        call.enqueue(object : Callback<WeatherResponse?> {
            override fun onResponse(call: Call<WeatherResponse?>, response: Response<WeatherResponse?>) {

                if (response.isSuccessful) {
                    val res = response.body()
                    if (res != null) {
                        parseHourlyWeather(res.hourly)
                        parseCurrentWeather(res.daily)
                        parseDailyForecast(res.daily)
                    }
                } else {
                    //TODO Implement LocalCache
                    log("Something wrong => ${response.errorBody()}")
                }
            }

            override fun onFailure(call: Call<WeatherResponse?>, t: Throwable) {
                application.toast("Couldn't retrieve updated weather")
                t.printStackTrace()
            }
        })
    }

    fun parseHourlyWeather(listHourly: List<WeatherItem>) {
        val mList = mutableListOf<HourlyWeatherItem>()
        // Just get 24 hours no more
        for(i in 0 until 24) {
            val wItem = listHourly[i]
            val weather = wItem.weather[0]
            val resIcon = IconMapsUtil.mapIcon(weather.icon)
            val temp = String.format(Locale.US, "%.0f", wItem.temp)
            mList.add(HourlyWeatherItem(wItem.dt, temp, resIcon))
        }
        hourlyList.postValue(mList)
    }

    fun parseDailyForecast(listDaily: List<WeatherDailyResponse>) {
        val map = IconMapsUtil.weatherIconMap
        val mList = mutableListOf<DailyWeatherItem>()
        log("Daily size: ${listDaily.size}")

        for (i in 1 until 7) {
            val res = listDaily[i]

            val weather = res.weather[0]
            val day = dayFormat.format(Date(res.dateTime))
            val title = weather.title
            val temp = String.format(Locale.US, "%.0f", res.temp.day)
            val resIcon = IconMapsUtil.mapIcon(weather.icon)
            mList.add(DailyWeatherItem(res.dateTime, temp, title, resIcon))
        }
        dailyList.postValue(mList)
    }

    fun parseCurrentWeather(listDaily: List<WeatherDailyResponse>) {

        val dailyResponse = listDaily[0]
        val weather = dailyResponse.weather[0]


        val d = Date()
        val dayOfTheWeek = dayFormat.format(d)

        val title = weather.title
        val temperature = String.format(Locale.US, "%.0f", dailyResponse.temp.day)
        val humidity = String.format(Locale.US, "%.0f", dailyResponse.humidity) + "%"

        // Default metric (meter/s) and imperial (miles/h)
        var windValue = dailyResponse.wind
        // Miles/h to KM/h
        weatherUnit?.let {
            windValue *= if (it.contentEquals(IMPERIAL)) {
                1.60934
            } else {
                // Meter/Sec to KM/H
                3.6
            }
        }

        val windPressure = String.format(Locale.US, "%.0f km/h", windValue)

        val resIcon = IconMapsUtil.mapIcon(weather.icon)
        val currentWeatherItem =
                CurrentWeatherItem(dayOfTheWeek, temperature, title, humidity, windPressure, resIcon)

        currentWeather.postValue(currentWeatherItem)
    }
}